Handle overflow in SelectNext

acelaya opened the pull request:
> Closes #1249 

This PR is trying to make `SelectNext` behave closer to the native `select` when the items overflow the size of the select itself.

For context, this is how a native select looks in that case:

![Captura desde 2023-10-16 15-33-07](https://github.com/hypothesis/frontend-shared/assets/2719332/106ff779-bd73-423c-adae-83fff7e09210)

* The content in the select itself is cropped, to make sure it does not overflow its container.
* The listbox grows to fit all options without cropping them. This is important, as there has to be a place where you can see the full content for every option.

Taking that into consideration, the next changes have been made.

1. The listbox has changed from `w-full` to `min-w-full`. That way it keeps looking the same as before in most of the cases, but it grows if the content is too long.

    ![image](https://github.com/hypothesis/frontend-shared/assets/2719332/e9ef6132-87d0-4ab3-bd96-5bc5771844d6)
    ![image](https://github.com/hypothesis/frontend-shared/assets/2719332/08647740-04ea-4fb3-8daf-37f9edd694a0)

2. The toggle button now truncates its content. That ensures there won't be content overflow, and displays ellipsis if the content is just plain text.
  When the content is not plain text, consumers might need to adapt their code in order to make it look nice. An example of this is included in the pattern library docs.

    ![image](https://github.com/hypothesis/frontend-shared/assets/2719332/b583b329-ace1-4aaa-8907-321f26be18c4)

### Testing steps.

You can check the new examples at the bottom of http://localhost:4001/input-select-next

robertknight commented on src/components/input/SelectNext.tsx:
```diff
@@ -194,7 +194,7 @@ function SelectMain<T>({
         variant="custom"
         classes={classnames(
-          'w-full flex',
+          'w-full flex justify-between',
```
What is the `justify-between` for here? To replace the `grow` spacer below?

acelaya replied:
Yes. The `text-overflow: ellipsis` didn't work on a `grow` element.

robertknight commented on src/pattern-library/components/patterns/prototype/SelectNextPage.tsx:
```diff
@@ -216,11 +222,31 @@ export default function SelectNextPage() {
+            </p>
+
+            <Library.Demo title="Plain text">
+              <div className="mx-auto">
```
The width and left position of the select control changes depending on the selected item. That looks weird to me, and it means that the dropdown arrow moves around. I think the widths of these controls is almost always fixed.

acelaya replied:
> The width and left position of the select control changes depending on the selected item

This is precisely what I was trying to prevent. I think I may have broken in some of my latest changes before pushing the branch.

It should behave like in the second screenshot I added to the PR description. I'll try to figure it out.

acelaya replied:
Yeah, I see the problem.

In order for this to work, the dimensions need to be somehow constrained, either by the `SelectNext` itself (like here) or its container. Otherwise, it can potentially grow.

In this case, since I'm trying to constrain it through `SelectNext`'s own `classes` prop, it needs to be `!w-36`, not `w-36`, otherwise the `w-full` it already has will take precedence.

Another option would be to constrain it through the container (I think that's what I did initially, when I took the screenshots), but it's important the other option also works, in case we need to use it inside an `InputGroup`.

Having to add the `important` modifier is not ideal (it means you need to know the internal details), but we can think on a better way to address that later.

robertknight commented on src/pattern-library/components/patterns/prototype/SelectNextPage.tsx:
```diff
@@ -216,11 +222,31 @@ export default function SelectNextPage() {
+
+          <Library.Example title="Select with long content">
+            <p>
+              <code>SelectNext</code> makes sure the content never overflows,
```
Are you referring to the button or listbox content here?

acelaya replied:
I think I wrote this at one point when I had a different approach in mind than what I ended up pushing 😅 

I'll re-write it to make it more clear.

robertknight commented on src/pattern-library/components/patterns/prototype/SelectNextPage.tsx:
```diff
@@ -216,11 +222,31 @@ export default function SelectNextPage() {
+            <p>
+              <code>SelectNext</code> makes sure the content never overflows,
+              but you may need to tweak <code>buttonContent</code> or options
+              content in cases where more complex content needs to be rendered.
```
What kind of "tweak" do you mean?

acelaya replied:
I'll try to clarify it.

Basically, the toggle button provides some affordances in the form of `text-overflow: ellipsis` and `overflow: hidden`.

While the later is always going to apply no matter the type of children provided to `buttonContent`, the former only applies when it is "plain text" (so to speak).

If you want a more complex set of children to rearrange differently, it's up to the consumer to handle it.

For example, I have included an example in the pattern library where the children have its own `truncate` wrapper to make sure the badge with the number is always displayed and only the text is hidden.

What I was trying to convey is that every use case might be different and it's up to downstream projects to customize it if the default `overflow: hidden` is not right for them.

acelaya commented:
@robertknight I have pushed the next changes.

* I have tried to improve the explanation of the component's behavior when content would overflow, trying to clarify the behaviors of the toggle button vs the listbox.
* I have fixed the miss-behavior you detected, where the button size changed depending on selected option, rather than hiding the overflowing content.
* I have added one more example where an `InputGroup` is used.

robertknight reviewed (approved):
LGTM
